### ISSUE TITLE
2.12-ification

### ISSUE THREAD

**Issue by mosesn** [twitter/finagle#505]
In order to support 2.12, we need to add crossScalaVersions to Build.scala, and make sure everything compiles.  Not every scala library will be built against the current snapshot, so we may have to support different versions from 2.11 to 2.12.  If part of an API has churned (scalacheck in particular is susceptible to this) we may need to upgrade the 2.11 library version to make progress.  If that happens, contact me (@mosesn) and I'll navigate the upgrade for you–it's a little complicated because we will need to upgrade all of our internal libraries, not just finagle.  Please note that most of this upgrade will not be feasible until we have a portion of util already upgraded, and we won't be able to do the thrifty bits until scrooge is upgraded too.
- [x] finagle-benchmark
- [x] finagle-benchmark-thrift
- [x] finagle-cacheresolver
- [x] finagle-commons-stats
- [x] finagle-core
- [x] finagle-example
- [x] finagle-exception
- [x] finagle-exp
- [x] finagle-http
- [x] finagle-http-compat
- [x] finagle-http-netty4
- [x] finagle-http2
- [x] finagle-integration
- [x] finagle-kestrel
- [x] finagle-mdns
- [x] finagle-memcached
- [x] finagle-mux
- [x] finagle-mysql
- [x] finagle-native
- [x] finagle-netty4
- [x] finagle-netty4-http
- [x] finagle-ostrich4
- [x] finagle-redis
- [x] finagle-serversets
- [x] finagle-spdy
- [x] finagle-stats
- [x] finagle-stream
- [x] finagle-stress
- [x] finagle-test
- [x] finagle-testers
- [x] finagle-thrift
- [x] finagle-thriftmux
- [x] finagle-zipkin

If you want to claim a module, please let us know in this thread!  I'll annotate different modules once they're done.

Probably a good way to start is for someone to try add 2.12 support for all of them, and then report back which ones still fail, so we know which ones are tricky and which ones work trivially.

**Comment by taylorleese** [comment]
Related:
- https://github.com/twitter/util/issues/162
- https://github.com/twitter/scrooge/issues/236

**Comment by jeffreyolchovy** [comment]
@mosesn let me know if you still want any outside help with this. when i locally hacked around with ostrich to complete my dependencies to attempt a trial 2.12 upgrade for finagle there didn't seem to be many necessary *finagle source* changes.

off the top of my head, iirc, most things centered around dependencies:
- upgrade jackson version to one that supports 2.12 (i.e. for jackson scala module)
- twitter bijection needs to support 2.12 for finagle-memcached

however, the main issue that i noted was compilation of 2.12 barks with an ambiguous reference to overloaded definition for the compose operators on `Filter`:
```
[error] both method andThen in class Filter of type (f: Req => com.twitter.util.Future[Rep])Req => com.twitter.util.Future[Rep]
[error] and  method andThen in class Filter of type (service: com.twitter.finagle.Service[Req,Rep])com.twitter.finagle.Service[Req,Rep]
[error] match argument types (com.twitter.finagle.Service[Req,Rep]) and expected result type com.twitter.finagle.Service[Req,Rep]
[error]         service => filters.andThen(service)
```

hopefully, resolving this one will not need an API change, but I think decisions like this are best left to you guys :)

anyway, this was back at the end of May, so maybe all is moot and a non-issue now, but like i said, let me know if you'd like an outside set of hands on this.

**Comment by mosesn** [comment]
@jeffreyolchovy yeah, we ran into that too.  filed a ticket in the scala issue tracker here: https://issues.scala-lang.org/browse/SI-9871 and it's being worked on here: https://github.com/scala/scala-dev/issues/157

we'd love some help with poking the bijection people to start supporting 2.12–I don't know if any of them are still at Twitter.  jackson we're waiting on the same thing as in the ostrich ticket.

good to hear it won't require any source changes in finagle!  and thanks so much for digging into this.

**Comment by noctella** [comment]
We've done some further poking of the bijection folks; once that's in I think we'll be able to start 2.12-ification in earnest. @jeffreyolchovy are you still interested?

**Comment by jcrossley** [comment]
We've done some further poking of the bijection folks; once that's in I think we'll be able to start 2.12-ification in earnest. @jeffreyolchovy are you still interested?

**Comment by jeffreyolchovy** [comment]
Sure. Myself and a few other @Tapad engineers would like to contribute toward the effort :)

**Comment by jcrossley** [comment]
@jeffreyolchovy that's great to hear, thank you! Will let you know once the bijection upgrade lands.

**Comment by clhodapp** [comment]
Bijection 0.9.3 with support for 2.12 is on maven central.

**Comment by jeffreyolchovy** [comment]
I'll start revisiting this weekend so that our team members are ready to contribute early next week.

**Comment by mosesn** [comment]
@jeffreyolchovy bad news, there's a scala bug that's preventing us from upgrading now.  https://issues.scala-lang.org/browse/SI-9871

I'm not sure what we can do until that gets resolved.  @SethTisue do you know if there's anything we can do to work around this in the mean time?

**Comment by jeffreyolchovy** [comment]
Ah yeah.. Definitely a blocker unless you want to refactor a bunch of core code...

**Comment by mosesn** [comment]
Update on this–we ended up having to make a pretty painful breaking change in finagle to work around the scalac backwards incompatibility, since the scala team introduced new semantics in 2.12 that unbreaking our backwards incompatibility would have made backwards incompatible.  Subsequently, we ran into several more compiler bugs, so @jcrossley ended up tackling it, and has migrated finagle to 2.12!  Yay!!  Marking this one as closed.  Thanks for your help, @jeffreyolchovy!
